submissions/Chikus/port-sniffer/sniffer.js

```diff
@@ -0,0 +1,123 @@
+const Net = require('net');
+
+function sniffer(arg) {
+  if (arg.length > 6) {
+    throw new Error(
+      'Bad usage to know the manual type: node sniffer.js --help'
+    );
+  }
+  const help = `
+TCP sniffer parameters:
+    --host     Mandatory parameter which define the host address for a TCP
+               port, It can be specified as IP address or URL.
+    --ports    Optional parameter which specifies port range, from min (1) to
+               max (65536) with a separator '-' between them, if this is not specified will
+               take the min and max values.
+
+Examples: node sniffer.js --host google.com --ports 100-2345
+          node sniffer.js --ports 1-1500 --host www.google.com
+          node sniffer.js --host 171.217.3.110
+          node sniffer.js --ports 70-80 --host 171.217.3.110
+          node sniffer.js --host www.google.com
+`;
+
+  function checkServer(host, port) {
+    return new Promise(function prom(resolve) {
+      const socket = Net.createConnection(port, host);
+      socket.setTimeout(300);
+      socket.on('timeout', function timer() {
+        socket.destroy();
+        resolve(null);
+      });
+      socket.on('connect', function success() {
+        process.stdout.write('.');
+        socket.destroy();
+        resolve(port);
+      });
+      socket.on('error', function fail() {
+        throw new Error('Error in Socket, wrong host \n');
+      });
+    }).catch(e => e);
+  }
+
+  async function runPorts(host, lowerPort, higherPort) {
+    const portList = [];
+    for (let i = lowerPort; i <= higherPort; i += 1) {
+      /* eslint-disable */
+      const validPort = await checkServer(host, i);
+      /* eslint-enable */
+      if (validPort) {
+        portList.push(validPort);
+      }
+    }
+    if (portList.length) {
+      process.stdout.write(`\n${portList.join()} ports are opened \n`);
+      return process.exit(0);
+    }
+    process.stdout.write('No ports were found \n');
+    return process.exit(0);
+  }
+
+  function checkPorts(userPorts) {
+    if (userPorts) {
+      const ports = userPorts.split('-').map(elem => Number(elem));
+      if (
+        ports[0] > ports[1] ||
+        ports[0] < 0 ||
+        ports[1] > 65535 ||
+        !ports[1]
+      ) {
+        throw new Error(' Check your ports, range from 1-65535\n');
+      }
+      return [ports[0], ports[1]];
+    }
+    throw new Error('Ports no defined \n');
+  }
+
+  function checkHost(host) {
+    if (host) {
+      const domains = host.split('.');
```

Comment:
```js
const dns = require('dns').promises;
async function checkHost(host) {
    try {
       return (await dns.lookup(host)).address;
    } catch (err) {
      throw new Error('Wrong host');
    }
}
```

Reply:
Cool and Done

Reply:
I used this function but I consider it is not necessary to add, for example if you set in argument --host 10.4 this function will not throw any error, it will fill with zeros the IP 10.0.0.4, also I had some issues with my node version, but when I maked it work, I see it is doing the same that net.socket does, so for this consideration I didn't implemented.

Reply:
DNS is not the same as `socket`. Socket uses TCP connection or IPC streaming. DNS is domain resolve system built on top of UDP with fallback to TCP.
If you don't want to use it - it's ok, but your current host validation is not correct and looks ugly. Here example of correct (RFC) regexp for hostname (https://stackoverflow.com/questions/106179/regular-expression-to-match-dns-hostname-or-ip-address). As you see its pretty big, you may use this if you want or check for Invalid host error in `socket.error` event to break the loop on first connect to wrong host.

Reply:
You are right, Socket doesnt do the same as DNS, but what I tried to said is that socket 
 = required('net')  inside of socket they are doing DNS validation, You can checked in this repo. https://github.com/nodejs/node/blob/master/lib/net.js , other wise how socket can work when you provide www.google.com here im sure they first contact DNS to get the IP add.

Reply:
Yes, it does. Generally, all you need is to check host validity once, at the start of your job.
I think that using `dns` for this purpose appears to be more clear then doing it through socket.connect. Its like using a bicycle  when you need a wheel.